### PR TITLE
add python 3.11 support

### PR DESCRIPTION
**Pull request by charles-cooper**
### Commit message

Commit message for the final, squashed PR. (Optional, but reviewers will appreciate it! Please see [our commit message style guide](../../blob/master/docs/style-guide.rst#best-practices-1) for what we would ideally like to see in a commit message.)

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

**Comment by charles-cooper** [comment]
waiting on https://github.com/actions/runner-images/issues/6459

**Comment by chanhosuh** [comment]
I'm a bit eager to get the 3.11 support.. is everything good here? :)

**Comment by charles-cooper** [comment]
> I'm a bit eager to get the 3.11 support.. is everything good here? :)

we are still waiting on gh actions runners to support

**Comment by chanhosuh** [comment]
> waiting on [actions/runner-images#6459](https://github.com/actions/runner-images/issues/6459)

if I read your guys' [test yaml](https://github.com/vyperlang/vyper/blob/master/.github/workflows/test.yml) righ, you need python 3.11 in this [runner image](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md).  It's listed there now!

**Comment by charles-cooper** [comment]
now waiting on https://github.com/ethereum/web3.py/pull/2699

File: setup.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def _global_version(version):
     keywords="ethereum evm smart contract language",
     include_package_data=True,
     packages=find_packages(exclude=("tests", "docs")),
-    python_requires=">=3.7,<3.11",
+    python_requires=">=3.8,<4",
```
Comment (fubuloubu) [inline setup.py:92]: Not sure when we can support 3.12, but probably not out of the gate
```suggestion
    python_requires=">=3.8,<3.12",
```

**Review by fubuloubu — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (charles-cooper) [reply setup.py:92]: i think we should leave the upper bound unspecified so if/when the next python is released, people can try using vyper without us needing to explicitly open a PR and testing it

Comment (fubuloubu) [reply setup.py:92]: Okay. I suppose the metadata shows that we don't explicitly support it yet

**Review by fubuloubu — APPROVED** [review]

**Comment by chanhosuh** [comment]
Seems like upgrading `web3` to the latest beta should do it (v6.0.0-beta.10): https://github.com/ethereum/web3.py/pull/2699#issuecomment-1431829014

**Comment by charles-cooper** [comment]
Will check! Note that since web3 is only required for our test suite- as of https://github.com/vyperlang/vyper/pull/3195 you should be able to install vyper with 3.11, it's just not tested/ officially supported.

File: setup.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
         "pytest-rerunfailures>=10.2,<11",
         "eth-tester[py-evm]>=0.6.0b6,<0.7",
         "py-evm>=0.5.0a3,<0.6",
-        "web3==5.27.0",
+        "web3==5.31.3",
```
Comment (fubuloubu) [inline setup.py:19]: Can probably keep it like this just for occasionally testing updates for v6 until it's released
```suggestion
        "web3>=6.0.0b10,<7.0.0",
```

**Pull request merged**